File: TestDPLS-iOS/TestDPLS/Protocol/DplsProtocol.swift
import Foundation

enum DplsProtocol {
    static let version: UInt8 = 1
    static let headerSize = 7
    static let trailerSize = 2
    static let overhead = headerSize + trailerSize

    enum MessageType: UInt8 {
        case hello = 0x01
        case authChallenge = 0x02
        case authProof = 0x03
        case authResult = 0x04
        case setup = 0x05
        case deviceInfoGet = 0x06
        case deviceInfoReport = 0x07
        case nameSet = 0x08
        case passwordSet = 0x09
        case settingsResult = 0x0a
        case stateGet = 0x10
        case stateReport = 0x11
        case modeSet = 0x12
        case commandResult = 0x13
        case identifyStart = 0x14
        case identifyStop = 0x15
        case logStart = 0x20
        case logInfo = 0x21
        case logChunk = 0x22
        case logAck = 0x23
        case logFinish = 0x24
        case logResult = 0x25
        case keepAlive = 0x30
        case error = 0x7f
    }

    struct Frame: Equatable {
        let type: MessageType
        let sequence: UInt16
        var flags: UInt8 = 0
        var payload: Data = Data()
    }

    enum DecodeResult: Equatable {
        case success(Frame)
        case failure(String)
    }

    static func encode(_ frame: Frame) -> Data {
        precondition(frame.payload.count <= 0xffff)
        var bytes = Data(capacity: overhead + frame.payload.count)
        bytes.append(version)
        bytes.append(frame.type.rawValue)
        bytes.append(frame.flags)
        bytes.append(contentsOf: withUnsafeBytes(of: frame.sequence.littleEndian, Array.init))
        bytes.append(contentsOf: withUnsafeBytes(of: UInt16(frame.payload.count).littleEndian, Array.init))
        bytes.append(frame.payload)
        let crc = crc16(bytes)
        bytes.append(contentsOf: withUnsafeBytes(of: UInt16(crc).littleEndian, Array.init))
        return bytes
    }

    static func decode(_ bytes: Data) -> DecodeResult {
        guard bytes.count >= overhead else { return .failure("Короткий кадр") }
        guard bytes[0] == version else { return .failure("Версия протокола не поддерживается") }
        guard let type = MessageType(rawValue: bytes[1]) else { return .failure("Неизвестный тип сообщения") }
        let flags = bytes[2]
        let sequence = readU16(bytes, at: 3)
        let payloadLength = Int(readU16(bytes, at: 5))
        guard bytes.count == overhead + payloadLength else { return .failure("Неверная длина кадра") }
        let expected = Int(readU16(bytes, at: bytes.count - 2))
        let actual = crc16(bytes.prefix(bytes.count - 2))
        guard expected == actual else { return .failure("Ошибка CRC16") }
        let payload = bytes.subdata(in: headerSize ..< headerSize + payloadLength)
        return .success(Frame(type: type, sequence: sequence, flags: flags, payload: payload))
    }

    /// CRC-16/CCITT-FALSE (init 0xFFFF, poly 0x1021, no reflection) — same as firmware.
    static func crc16(_ bytes: Data) -> Int {
        var crc = 0xffff
        for byte in bytes {
            crc ^= Int(byte) << 8
            for _ in 0 ..< 8 {
                crc = (crc & 0x8000) != 0 ? ((crc << 1) ^ 0x1021) : (crc << 1)
                crc &= 0xffff
            }
        }
        return crc
    }

    private static func readU16(_ data: Data, at offset: Int) -> UInt16 {
        UInt16(data[offset]) | (UInt16(data[offset + 1]) << 8)
    }
}

enum LittleEndian {
    static func u8(_ data: Data, _ offset: inout Int) -> UInt8 {
        defer { offset += 1 }
        return data[offset]
    }

    static func u16(_ data: Data, _ offset: inout Int) -> UInt16 {
        defer { offset += 2 }
        return UInt16(data[offset]) | (UInt16(data[offset + 1]) << 8)
    }

    static func u32(_ data: Data, _ offset: inout Int) -> UInt32 {
        defer { offset += 4 }
        return UInt32(data[offset])
            | (UInt32(data[offset + 1]) << 8)
            | (UInt32(data[offset + 2]) << 16)
            | (UInt32(data[offset + 3]) << 24)
    }

    static func appendU8(_ data: inout Data, _ value: UInt8) { data.append(value) }
    static func appendU16(_ data: inout Data, _ value: UInt16) {
        data.append(contentsOf: withUnsafeBytes(of: value.littleEndian, Array.init))
    }
    static func appendU32(_ data: inout Data, _ value: UInt32) {
        data.append(contentsOf: withUnsafeBytes(of: value.littleEndian, Array.init))
    }
}

File: TestDPLS-Windows/src/TestDPLS.Core/Protocol/DplsProtocol.cs
using System.Buffers.Binary;

namespace TestDPLS.Protocol;

/// <summary>
/// Wire framing shared with firmware, Android and iOS clients.
/// CRC-16/CCITT-FALSE (init 0xFFFF, poly 0x1021, no reflection).
/// </summary>
public static class DplsProtocol
{
    public const byte Version = 0x01;
    public const int HeaderSize = 7;
    public const int TrailerSize = 2;
    public const int Overhead = HeaderSize + TrailerSize;

    public enum MessageType : byte
    {
        Hello = 0x01,
        AuthChallenge = 0x02,
        AuthProof = 0x03,
        AuthResult = 0x04,
        Setup = 0x05,
        DeviceInfoGet = 0x06,
        DeviceInfoReport = 0x07,
        NameSet = 0x08,
        PasswordSet = 0x09,
        SettingsResult = 0x0a,
        StateGet = 0x10,
        StateReport = 0x11,
        ModeSet = 0x12,
        CommandResult = 0x13,
        IdentifyStart = 0x14,
        IdentifyStop = 0x15,
        LogStart = 0x20,
        LogInfo = 0x21,
        LogChunk = 0x22,
        LogAck = 0x23,
        LogFinish = 0x24,
        LogResult = 0x25,
        KeepAlive = 0x30,
        Error = 0x7f,
    }

    public readonly record struct Frame(MessageType Type, ushort Sequence, byte[] Payload, byte Flags = 0);

    public abstract record DecodeResult
    {
        public sealed record Success(Frame Frame) : DecodeResult;
        public sealed record Failure(string Reason) : DecodeResult;
    }

    public static byte[] Encode(Frame frame)
    {
        if (frame.Payload.Length > 0xffff)
            throw new ArgumentOutOfRangeException(nameof(frame), "Payload too large");

        var bytes = new byte[Overhead + frame.Payload.Length];
        bytes[0] = Version;
        bytes[1] = (byte)frame.Type;
        bytes[2] = frame.Flags;
        BinaryPrimitives.WriteUInt16LittleEndian(bytes.AsSpan(3), frame.Sequence);
        BinaryPrimitives.WriteUInt16LittleEndian(bytes.AsSpan(5), (ushort)frame.Payload.Length);
        frame.Payload.CopyTo(bytes, HeaderSize);
        var crc = Crc16(bytes.AsSpan(0, HeaderSize + frame.Payload.Length));
        BinaryPrimitives.WriteUInt16LittleEndian(bytes.AsSpan(HeaderSize + frame.Payload.Length), crc);
        return bytes;
    }

    public static DecodeResult Decode(ReadOnlySpan<byte> bytes)
    {
        if (bytes.Length < Overhead)
            return new DecodeResult.Failure("Короткий кадр");
        if (bytes[0] != Version)
            return new DecodeResult.Failure("Версия протокола не поддерживается");
        if (!Enum.IsDefined(typeof(MessageType), bytes[1]))
            return new DecodeResult.Failure("Неизвестный тип сообщения");

        var type = (MessageType)bytes[1];
        var flags = bytes[2];
        var sequence = BinaryPrimitives.ReadUInt16LittleEndian(bytes.Slice(3));
        var payloadLength = BinaryPrimitives.ReadUInt16LittleEndian(bytes.Slice(5));
        if (bytes.Length != Overhead + payloadLength)
            return new DecodeResult.Failure("Неверная длина кадра");

        var expected = BinaryPrimitives.ReadUInt16LittleEndian(bytes.Slice(bytes.Length - 2));
        var actual = Crc16(bytes.Slice(0, bytes.Length - 2));
        if (expected != actual)
            return new DecodeResult.Failure("Ошибка CRC16");

        var payload = bytes.Slice(HeaderSize, payloadLength).ToArray();
        return new DecodeResult.Success(new Frame(type, sequence, payload, flags));
    }

    /// <summary>CRC-16/CCITT-FALSE — same as firmware.</summary>
    public static ushort Crc16(ReadOnlySpan<byte> bytes)
    {
        ushort crc = 0xffff;
        foreach (var b in bytes)
        {
            crc ^= (ushort)(b << 8);
            for (var i = 0; i < 8; i++)
                crc = (crc & 0x8000) != 0 ? (ushort)((crc << 1) ^ 0x1021) : (ushort)(crc << 1);
        }
        return crc;
    }
}

public static class LittleEndian
{
    public static byte U8(ReadOnlySpan<byte> data, ref int offset)
    {
        var v = data[offset];
        offset += 1;
        return v;
    }

    public static ushort U16(ReadOnlySpan<byte> data, ref int offset)
    {
        var v = BinaryPrimitives.ReadUInt16LittleEndian(data.Slice(offset));
        offset += 2;
        return v;
    }

    public static uint U32(ReadOnlySpan<byte> data, ref int offset)
    {
        var v = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset));
        offset += 4;
        return v;
    }

    public static void AppendU8(List<byte> data, byte value) => data.Add(value);

    public static void AppendU16(List<byte> data, ushort value)
    {
        Span<byte> tmp = stackalloc byte[2];
        BinaryPrimitives.WriteUInt16LittleEndian(tmp, value);
        data.AddRange(tmp.ToArray());
    }

    public static void AppendU32(List<byte> data, uint value)
    {
        Span<byte> tmp = stackalloc byte[4];
        BinaryPrimitives.WriteUInt32LittleEndian(tmp, value);
        data.AddRange(tmp.ToArray());
    }

    public static byte[] Concat(params byte[][] parts)
    {
        var len = 0;
        foreach (var p in parts) len += p.Length;
        var result = new byte[len];
        var o = 0;
        foreach (var p in parts)
        {
            p.CopyTo(result, o);
            o += p.Length;
        }
        return result;
    }
}
